test/aoc.test.js

```diff
@@ -87,6 +87,65 @@ let res = func({input, udf})
 expect(res).toBe(281)
 })
 
+test("day2", () => {
+  // used udfs to simulate behaviors of select, all, scan in jq
+  let input = `Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
+Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
+Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
+Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
+Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green`
+
+  let bag = {
+    red: 12,
+    green: 13,
+    blue: 14
+  }
+  
+  let udf = {
+    splitN: x => x.split("\n"),
+    splitColon: x => x.split(":"),
+    splitSpace: x => x.split(" "),
+    splitSemicolon: x => x.split(";"),
+    splitComma: x => x.split(","),
+    // TODO: try to get rid of all()
+    all: x => x.every(v => v) ? 1 : 0,
+    isNonNegative: x => x >= 0,
+    toNum: x => Number(x)
+  }
+  
+  let lines = rh`input | udf.splitN | *line
+                       | udf.splitColon | *part`
+
+  let id = rh`.input | udf.splitN | .*line
+                     | udf.splitColon | .*part | first
+                     | udf.splitSpace | .*id | last
+                     | udf.toNum`
+  
+  let nums = rh`.input | udf.splitN | .*line
+                       | udf.splitColon | .*part | last
+                       | udf.splitSemicolon | .*hand
+                       | udf.splitComma | .*group
+                       | udf.splitSpace | .1
+                       | udf.toNum`
+
+  let colors = rh`.input | udf.splitN | .*line
+                         | udf.splitColon | .*part | last
+                         | udf.splitSemicolon | .*hand
+                         | udf.splitComma | .*group
+                         | udf.splitSpace | .2`
+
+  let diff = api.apply("udf.isNonNegative", api.minus(api.get(".bag", colors), nums))
+
+  let lineRes = api.times(id, api.apply("udf.all", api.array(diff)))
+
+  let query = pipe(lineRes).group("*line").get("*").sum()
+
```

Comment:
How about using the textual frontend for these as well?

Reply:
Sure, I will look into that.